pkg/planner/core/logical_plan_builder.go

```diff
@@ -5266,76 +5268,184 @@ func buildColumns2HandleWithWrtiableColumns(
+	// If there is foreign key, we can't prune the columns.
+	if !hasFK {
+		nonPruned = bitset.New(uint(len(names)))
+		nonPruned.SetAll()
+	}
+	cols2PosInfos := make(TblColPosInfoSlice, 0, len(tblID2Handle))
+	for tid, handleCols := range tblID2Handle {
 		for _, handleCol := range handleCols {
-			curColPosInfo, err := buildSingleTableColPosInfoForDelete(names, handleCol, deletableIdxs, deletableCols, tblInfo)
+			curColPosInfo, err := initColPosInfo(tid, names, handleCol)
 			if err != nil {
-				return nil, err
+				return nil, nil, err
 			}
 			cols2PosInfos = append(cols2PosInfos, curColPosInfo)
 		}
 	}
-	sort.Sort(cols2PosInfos)
-	return cols2PosInfos, nil
+	// Sort by start position. To do the later column pruning.
+	slices.SortFunc(cols2PosInfos, func(i, j TblColPosInfo) int {
+		if i.Start < j.Start {
+			return -1
+		}
+		if i.Start > j.Start {
+			return 1
+		}
+		return 0
+	})
+	prunedColCnt := 0
+	var err error
+	for i := range cols2PosInfos {
+		cols2PosInfo := &cols2PosInfos[i]
+		tbl := tblID2Table[cols2PosInfo.TblID]
+		deletableIdxs := tbl.DeletableIndices()
+		deletableCols := tbl.DeletableCols()
+		tblInfo := tbl.Meta()
+		publicCols := tbl.Cols()
+		prunedColCnt, err = pruneAndBuildSingleTableColPosInfoForDelete(names, deletableIdxs, deletableCols, publicCols, tblInfo, cols2PosInfo, prunedColCnt, nonPruned)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return cols2PosInfos, nonPruned, nil
+}
+
+func initColPosInfo(tid int64, names []*types.FieldName, handleCol util.HandleCols) (TblColPosInfo, error) {
+	offset, err := getTableOffset(names, names[handleCol.GetCol(0).Index])
+	if err != nil {
+		return TblColPosInfo{}, err
+	}
+	return TblColPosInfo{
+		TblID:                 tid,
+		Start:                 offset,
+		HandleCols:            handleCol,
+		ExtraPartialRowOption: &table.ExtraPartialRowOption{},
+	}, nil
 }
 
-// buildSingleTableColPosInfoForDelete builds columns mapping for delete.
-func buildSingleTableColPosInfoForDelete(
+// pruneAndBuildSingleTableColPosInfoForDelete builds columns mapping for delete.
+// And it will try to prune columns that not used in pk or indexes.
+func pruneAndBuildSingleTableColPosInfoForDelete(
 	names []*types.FieldName,
-	handleCol util.HandleCols,
 	deletableIdxs []table.Index,
-	deletableCols []*table.Column,
+	deletableCols, publicCols []*table.Column,
 	tblInfo *model.TableInfo,
-) (TblColPosInfo, error) {
+	colPosInfo *TblColPosInfo,
+	prePrunedCount int,
+	nonPrunedSet *bitset.BitSet,
+) (int, error) {
 	// Columns can be seen by DELETE are the deletable columns.
 	tblLen := len(deletableCols)
-	offset, err := getTableOffset(names, names[handleCol.GetCol(0).Index])
-	if err != nil {
-		return TblColPosInfo{}, err
+
+	// If it's partitioned table, or has foreign keys, or is point get plan, we can't prune the columns, currently.
+	// nonPrunedSet will be nil if it's a point get or has foreign keys.
+	if tblInfo.GetPartitionInfo() != nil || nonPrunedSet == nil {
```

Comment:
Why are we using two different methods to pass the two similar kinds of information?

Reply:
And since this `if` block doesn't reuse any other logic in this function, why do we make this part a separate function? This function is already too large and lacks comments.